nixos/installer: Allow setting a password on cmdline for pxe boot

max06 opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
Copied isoImage-functionality for setting a password on boot for the nixos-user to the pxe image. 

The pxe images come without passwords for the `nixos` and `root` accounts. This works great when doing local installations, but it prevents you from using ssh into the live-system without setting a password first. Adding `live.nixos.passwd=somesecretpassword` to the cmdline in any pxe configuration eliminates this requirement. This change also eliminates the need to build own images to achive this goal.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

max06 committed:
feat: Allow setting a password on cmdline for pxe boot

max06 commented:
PR is caused by https://github.com/nix-community/nixos-images/issues/303

Mic92 commented on nixos/modules/installer/netboot/netboot.nix:
```diff
@@ -141,6 +141,17 @@ with lib;
+        # Allows using nixos-anywhere in headless environments
+        for o in $(</proc/cmdline); do
+          case "$o" in
+            live.nixos.passwd=*)
```
I was thinking about how to make this a bit more secure, netboot often is not encrypted and while we cannot protect against active attacker trying to intercept/change the traffic, we shouldn't have a service where we can just ask for the password using the netboot protocol. This is a bit different from the ISO usecase which is just local and requires physical access.
So our perl script currently supports setting password via password hashes like this:

```
foreach my $u (values %usersOut) {
    next if defined $shadowSeen{$u->{name}};
    my $hashedPassword = "!";
    $hashedPassword = $u->{hashedPassword} if defined $u->{hashedPassword};
    my $expires = "";
    $expires = dateToDays($u->{expires}) if defined $u->{expires};
    # FIXME: set correct value for sp_lstchg.
    push @shadowNew, join(":", $u->{name}, $hashedPassword, "1::::", $expires, "") . "\n";
}
```

Could we change this bash instead or awk and do something similar?

This is how to generate a password hash:

```
echo foo | mkpasswd -m sha-512 -s
```

Mic92 replied:
There is a difference between an attacker that can ask the PXE server for the same configuration and an attacker that pretends to be the PXE server (active).

Mic92 replied:
You said you add the option, but I can't see it. Did you forgot to push?

max06 replied:
An attacker altering the response still makes no difference if it's a password or a hash. Needless to say an attacker in the position to run a pxe server (he needs control over the dhcp server as well for that) doesn't need a nixos netboot image to do harm. But I get your point with the cleartext password in the nix store. 

No, not forgotten. Just adding the test scenario. Pushing my current state though for your feedback.

Mic92 commented:
It would be great if you could add the new parameter to `boot.kernelParams` here: https://github.com/NixOS/nixpkgs/blob/d4b1fcdbe64eedd4829db68b208ff0bf81e34b13/nixos/tests/boot.nix#L91

And than check `/etc/shadow` if it also contains the generated password.

max06 committed:
Add passwordHash option

Mic92 commented on nixos/modules/installer/netboot/netboot.nix:
```diff
@@ -141,6 +141,21 @@ with lib;
+          case "$o" in
+            live.nixos.passwordHash=*)
+              set -- $(IFS==; echo $o)
+              sed -i "s/nixos::/nixos:$2:/" /etc/shadow
```
I don't think `/etc/shadow` exists at this point because it's only created in the activation script. You will need an activation script as mentioned here: https://github.com/Mic92/sops-nix/blob/53c853fb1a7e4f25f68805ee25c83d5de18dc699/modules/sops/default.nix#L451

max06 replied:
It works for the iso image - why shouldn't it work here?

Mic92 replied:
Never mind. postBootCommands is run after the activation phase.

max06 replied:
Stop scaring me, please 😂

max06 commented:
> It would be great if you could add the new parameter to `boot.kernelParams` here:
> 
> https://github.com/NixOS/nixpkgs/blob/d4b1fcdbe64eedd4829db68b208ff0bf81e34b13/nixos/tests/boot.nix#L91

I'm not sure if that even works:

After adding `machine.succeed("cat /proc/cmdline; grep 'serial' /proc/cmdline")` to the test-script...

```
machine: must succeed: cat /proc/cmdline; grep 'serial' /proc/cmdline
machine: output: bzImage init=/nix/store/xskdcn0dpzslay1sz59gq8dmyjjdwh6g-nixos-system-nixos-25.05pre-git/init initrd=initrd console=ttyS0 console=tty0 panic=1 boot.panic_on_fail clocksource=acpi_pm loglevel=7 net.ifnames=0 
```

(In other words: I have no clue how to fix that.)

max06 committed:
Rebuild password update functionality, add tests

max06 committed:
Add tests

max06 commented:
@Mic92 I fixed the missing kernel parameters and added the `live.nixos.passwordHash`-Option as well for testing. I also replaced the first implementation with `sed` by `usermod`, less issues with characters in the password hash.

Please review and merge, if there's nothing else to do here.
Thanks alot!
